apoc.initializer.cypher didn't work as expexcted

## Guidelines

Please note that GitHub issues are only meant for bug reports/feature requests. If you have questions on how to use APOC, please ask on [the Neo4j Discussion Forum](https://community.neo4j.com/c/neo4j-graph-platform/procedures-apoc) instead of creating an issue here.

## Expected Behavior (Mandatory)

I added this line in neo4j.conf:
apoc.initializer.cypher=CALL apoc.log.info("Some logging info")
in the expectation of seeing the logging info displayed on screen or recorded in log file. 

## Actual Behavior (Mandatory)

Neo4j server failed to start up due to the errors below: 

Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -1
    at java.lang.String.substring(Unknown Source)
    at apoc.util.Util.subMap(Util.java:218)
    at apoc.ApocConfiguration.get(ApocConfiguration.java:52)
    at apoc.cypher.CypherInitializer.available(CypherInitializer.java:22)
    at org.neo4j.helpers.Listeners.notifySingleListener(Listeners.java:145)
    at org.neo4j.helpers.Listeners.notify(Listeners.java:101)
    at org.neo4j.kernel.availability.DatabaseAvailabilityGuard.fulfill(DatabaseAvailabilityGuard.java:90)
    at org.neo4j.kernel.availability.DatabaseAvailability.start(DatabaseAvailability.java:59)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:452)

## How to Reproduce the Problem

Add the configuration line above into neo4j.conf. 

### Simple Dataset (where it's possibile)

```
//Insert here a set of Cypher statements that helps us to reproduce the problem

```


### Steps (Mandatory)

  1.
  1.
  1.

## Screenshots (where it's possibile)

## Specifications (Mandatory)

apoc 3.5.0.2 on neo4j 3.5.5

### Versions

  - OS:
  - Neo4j:
  - Neo4j-Apoc:

--- a/src/main/java/apoc/cypher/CypherInitializer.java
+++ b/src/main/java/apoc/cypher/CypherInitializer.java
@@ -5,10 +5,14 @@ import org.neo4j.kernel.availability.AvailabilityListener;
 import org.neo4j.kernel.internal.GraphDatabaseAPI;
 import org.neo4j.logging.Log;
 
+import java.util.Collections;
+import java.util.Map;
 import java.util.SortedMap;
 import java.util.TreeMap;
 
 public class CypherInitializer implements AvailabilityListener {
+    public static final String INITIALIZER_CYPHER = "initializer.cypher";
+
     private final GraphDatabaseAPI db;
     private final Log userLog;
 
@@ -19,7 +23,16 @@ public class CypherInitializer implements AvailabilityListener {
 
     @Override
     public void available() {
-        SortedMap<String, Object> initializers = new TreeMap<>(ApocConfiguration.get("initializer.cypher"));
+        Map<String, Object> stringObjectMap;
+
+        String singleInitializer = ApocConfiguration.get(INITIALIZER_CYPHER, null);
+        if (singleInitializer != null) {
+            stringObjectMap = Collections.singletonMap("1", singleInitializer);
+        } else {
+            stringObjectMap = ApocConfiguration.get(INITIALIZER_CYPHER);
+        }
+
+        SortedMap<String, Object> initializers = new TreeMap<>(stringObjectMap);
         for (Object initializer: initializers.values()) {
             String query = initializer.toString();
             try {

--- a/src/main/java/apoc/cypher/CypherInitializer.java
+++ b/src/main/java/apoc/cypher/CypherInitializer.java
@@ -1,14 +1,14 @@
 package apoc.cypher;
 
 import apoc.ApocConfiguration;
+import org.neo4j.graphdb.DependencyResolver;
+import org.neo4j.internal.kernel.api.procs.ProcedureSignature;
 import org.neo4j.kernel.availability.AvailabilityListener;
+import org.neo4j.kernel.impl.proc.Procedures;
 import org.neo4j.kernel.internal.GraphDatabaseAPI;
 import org.neo4j.logging.Log;
 
-import java.util.Collections;
-import java.util.Map;
-import java.util.SortedMap;
-import java.util.TreeMap;
+import java.util.*;
 
 public class CypherInitializer implements AvailabilityListener {
     public static final String INITIALIZER_CYPHER = "initializer.cypher";
@@ -23,25 +23,49 @@ public class CypherInitializer implements AvailabilityListener {
 
     @Override
     public void available() {
-        Map<String, Object> stringObjectMap;
 
-        String singleInitializer = ApocConfiguration.get(INITIALIZER_CYPHER, null);
-        if (singleInitializer != null) {
-            stringObjectMap = Collections.singletonMap("1", singleInitializer);
-        } else {
-            stringObjectMap = ApocConfiguration.get(INITIALIZER_CYPHER);
-        }
+        // run initializers in a new thread
+        // we need to wait until apoc procs are registered
+        // unfortunately an AvailabilityListener is triggered before that
+        new Thread(() -> {
 
-        SortedMap<String, Object> initializers = new TreeMap<>(stringObjectMap);
-        for (Object initializer: initializers.values()) {
-            String query = initializer.toString();
+            awaitApocProceduresRegistered();
+
+            Map<String, Object> stringObjectMap;
+            String singleInitializer = ApocConfiguration.get(INITIALIZER_CYPHER, null);
+            if (singleInitializer != null) {
+                stringObjectMap = Collections.singletonMap("1", singleInitializer);
+            } else {
+                stringObjectMap = ApocConfiguration.get(INITIALIZER_CYPHER);
+            }
+
+            SortedMap<String, Object> initializers = new TreeMap<>(stringObjectMap);
+            for (Object initializer: initializers.values()) {
+                String query = initializer.toString();
+                try {
+                    db.execute(query);
+                    userLog.info("successfully initialized: " + query);
+                } catch (Exception e) {
+                    userLog.warn("error upon initialization, running: "+query, e);
+                }
+            }
+        }).start();
+    }
+
+    private void awaitApocProceduresRegistered() {
+        while (!areApocProceduresRegistered()) {
             try {
-                db.execute(query);
-                userLog.info("successfully initialized: " + query);
-            } catch (Exception e) {
-                userLog.warn("error upon initialization, running: "+query, e);
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
             }
         }
+    }
+
+    private boolean areApocProceduresRegistered() {
+        Procedures procs = db.getDependencyResolver().resolveDependency(Procedures.class, DependencyResolver.SelectionStrategy.FIRST);
+        Set<ProcedureSignature> allProcedures = procs.getAllProcedures();
+        return allProcedures.stream().anyMatch(signature -> signature.name().toString().startsWith("apoc"));
     }
 
     @Override

--- a/src/test/java/apoc/cypher/CypherInitializerTest.java
+++ b/src/test/java/apoc/cypher/CypherInitializerTest.java
@@ -17,16 +17,27 @@ public class CypherInitializerTest {
     public void init(String... initializers) {
         GraphDatabaseBuilder graphDatabaseBuilder = new TestGraphDatabaseFactory().newImpermanentDatabaseBuilder();
 
-        int index = 1;
-        for (String initializer: initializers) {
-            graphDatabaseBuilder.setConfig("apoc.initializer.cypher." + index++, initializer);
+        if (initializers.length == 1) {
+            graphDatabaseBuilder.setConfig("apoc.initializer.cypher", initializers[0]);
+        } else {
+            int index = 1;
+            for (String initializer: initializers) {
+                graphDatabaseBuilder.setConfig("apoc.initializer.cypher." + index++, initializer);
+            }
         }
+
         db = graphDatabaseBuilder.newGraphDatabase();
     }
 
     @After
     public void teardown() {
         db.shutdown();
+    }
+
+    @Test
+    public void noInitializerWorks() {
+        init();
+        expectNodeCount(0);
     }
 
     @Test

